Add support for Apple M1 AArch64 SoCs

arkivm opened the pull request:
Completely based on #150. Thanks to @sbehnke!
+ Refactoring to accomodate the new source tree
+ Adding more feature flags

As #150 is stale, I am just trying to push it across the finish line.

sbehnke reviewed (commented):
Awesome! Looks good to me. I'll have to give It a try as well. Sorry I was not able to get back to it. Life gets in the way.

arkivm commented:
No problem. Thanks for the great work. As numerous projects started tagging that they depend on this, I decided to push it forward.

arkivm commented:
As @sbehnke did all the work, can someone tell me how to add something like Linux's `Co-developed-by` tag to a github PR? I would like @sbehnke to sign off this patch and I can add myself as a co-developer.

sbehnke reviewed (approved):
```
➜  build git:(aarch64-darwin-support) make test 
Running tests...
Test project /Users/sbehnke/cpu_features/build
    Start 1: bit_utils_test
1/4 Test #1: bit_utils_test ...................   Passed    0.13 sec
    Start 2: string_view_test
2/4 Test #2: string_view_test .................   Passed    0.08 sec
    Start 3: stack_line_reader_test
3/4 Test #3: stack_line_reader_test ...........   Passed    0.07 sec
    Start 4: cpuinfo_aarch64_test
4/4 Test #4: cpuinfo_aarch64_test .............   Passed    0.07 sec

100% tests passed, 0 tests failed out of 4

Total Test time (real) =   0.35 sec
➜  build git:(aarch64-darwin-support) ./list_cpu_features 
arch            : aarch64
implementer     : 16777228 (0x100000C)
variant         :   2 (0x02)
part            : 458787763 (0x1B588BB3)
revision        :   5 (0x05)
flags           : aes,asimd,asimdfhm,atomics,crc32,fcma,fp,fphp,jscvt,lrcpc,pmull,sb,sha1,sha3,sha512,ssbs
```

Looks good on my M1 Max.

arkivm commented:
> Looks good on my M1 Max.

Great. Thank you for quickly testing it.

Mizux commented on CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.0)
+cmake_minimum_required(VERSION 3.19.2)
```
Why 3.19.2 ? this seems a little bit overkill can we limit it to max 3.16.3 (Ubuntu 20.04) or 3.10.2 (Ubuntu 18.04) ?

src: https://repology.org/project/cmake/badges

arkivm replied:
3.19.2 comes from native cmake host detection support : https://cmake.org/cmake/help/latest/release/3.19.html#platforms

Mizux replied:
erf my bad so if we want to correctly support M1 we'll need it...

arkivm replied:
yeah. I looked back that thread and it was you who posted this cmake link :)

Mizux commented on CMakeLists.txt:
```diff
@@ -52,10 +52,10 @@ set(PROCESSOR_IS_POWER FALSE)
 
 if(CMAKE_SYSTEM_PROCESSOR MATCHES "^mips")
   set(PROCESSOR_IS_MIPS TRUE)
+elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "(^aarch64)|(arm64)")
```
missing `^` in front of `arm64`

toor1245 commented on src/impl_aarch64__base_implementation.inl:
```diff
@@ -0,0 +1,100 @@
+// Copyright 2017 Google LLC
```
`impl_aarch64__base_implementation` implies that this is a general solution in which there may be a compiler dependency and there should be no OS dependency, like in `impl_x86__base_implementation`. Could you rename this file?

arkivm replied:
I removed all the OS dep from that file and retained the file name.

toor1245 commented on src/impl_aarch64_macos.c:
```diff
@@ -0,0 +1,81 @@
+}
+
+#endif  // CPU_FEATURES_OS_MACOS
+#endif  // CPU_FEATURES_ARCH_X86
```
aarch64...

toor1245 commented on src/impl_aarch64__base_implementation.inl:
```diff
@@ -0,0 +1,100 @@
+#include "cpu_features_macros.h"
+
+#ifdef CPU_FEATURES_ARCH_AARCH64
+#if defined(CPU_FEATURES_OS_LINUX) || defined(CPU_FEATURES_OS_ANDROID) || \
+    defined(CPU_FEATURES_OS_MACOS)
```
could there be an `CPU_FEATURES_OS_IPHONE` here?

arkivm replied:
I don't have the hardware to test. I have added this macro. is it possible to test it at your end?

sbehnke replied:
I can test on an aarch64 raspberry pi running Ubuntu, but don't have an easy setup to test Android.

toor1245 replied:
I can test the same only on aarch64 rpi Ubuntu

toor1245 replied:
I have verified that it works on linux, for iphone `sys/sysctl.h` it should theoretically work: https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/sysctl.3.html

arkivm replied:
thank you @toor1245 @sbehnke

toor1245 commented:
I don't see the `__arm64__` macro for `CPU_FEATURES_ARCH_AARCH64` in `cpu_features_macros.h`, could you provide?

toor1245 commented on src/impl_aarch64_macos.c:
```diff
@@ -0,0 +1,81 @@
+}
+
+static bool GetDarwinSysCtlByName(const char* name) {
+  return !!GetDarwinSysCtlByNameValue(name);
```
can be simplified to `GetDarwinSysCtlByNameValue(name) != 0`

toor1245 commented on src/impl_aarch64__base_implementation.inl:
```diff
@@ -0,0 +1,100 @@
+#include "internal/stack_line_reader.h"
+#include "internal/string_view.h"
+
+static bool HandleAarch64Line(const LineResult result, Aarch64Info* const info);
```
Clang-tidy: parameter is const-qualified in the function declaration; const-qualification of parameters only has an effect in function definitions

arkivm replied:
fixed. does it make sense to add clang-tidy to the CI?

toor1245 replied:
yes, I fully agree with you. @Mizux, @gchatelet, can we raise issue for this?

arkivm replied:
#207

toor1245 commented on src/impl_aarch64__base_implementation.inl:
```diff
@@ -0,0 +1,100 @@
+
+static bool HandleAarch64Line(const LineResult result, Aarch64Info* const info);
+
+static void FillProcCpuInfoData(Aarch64Info* const info);
```
same here

arkivm commented:
> I don't see the `__arm64__` macro for `CPU_FEATURES_ARCH_AARCH64` in `cpu_features_macros.h`? This macro is used for iphone

added this as well.

toor1245 commented on src/impl_aarch64__base_implementation.inl:
```diff
@@ -0,0 +1,93 @@
+// Implementation.
+////////////////////////////////////////////////////////////////////////////////
+
+static bool HandleAarch64Line(LineResult result, Aarch64Info* info);
+
+static void FillProcCpuInfoData(Aarch64Info* info);
```
`HandleAarch64Line` and `FillProcCpuInfoData` are Linux / Android only contract

arkivm replied:
If we `#ifdef` with `CPU_FEATURES_OS_LINUX || CPU_FEATURES_OS_ANDROID`, then this file would become dependent on the OS as opposed to only the compiler. so what do you suggest here? we can move these declarations to `aarch64_linux_or_android.c` or better remove the above declarations as it is not used anywhere else.

arkivm replied:
I did the latter (removed these decls). let me know if you want to have these in a common header.

toor1245 commented on src/impl_aarch64_macos.c:
```diff
@@ -0,0 +1,81 @@
+// Copyright 2021 Google LLC
```
rename `impl_aarch64_macos.c` to  `impl_aarch64_macos_or_iphone.c` like `impl_aarch64_linux_or_android.c`

arkivm commented:
bump

gchatelet commented:
Thx for the work here @arkivm we now have a clearer vision of what's needed to support Apple M1. That said I find the PR too big to be pulled in as-is. I'd rather implement it by creating many small PR that fixes individual issues (see #209 ). Let's keep the PR around for now while we fix it incrementally.

SchrodingerZhu commented:
How is the progress of merging?

toor1245 commented:
> How is the progress of merging?

Hi @SchrodingerZhu, No updates yet.

@gchatelet, since, there were no updates for a long time, can we discuss what components to divide? I can do it if I get help in testing @arkvim, @sbehnke. Plus a lot of things are already clear to take into account inline assembly https://github.com/google/cpu_features/pull/186.

mscdex commented:
Would it be possible to land this soon?

For what it's worth I've gone ahead and applied the changes (fixing several merge conflicts) on top of v0.8.0 and I have had at least one M1 user report it works (I do not have access to Apple Silicon hardware myself). If anyone is interested in these changes, they can be found in the patch [here](https://github.com/mscdex/cpu-features/blob/339bf44a0d5c5e7f255f6530875fbe2055cd9dd9/deps/cpu_features/patches/0001-Add-Apple-Silicon-Support.patch).

gchatelet commented:
I fixed the patch merging. @Mizux would you be able to test this patch on Apple silicon? Then it should be good to go.

Mizux commented:
FYI, I'm currently rebasing this branch on main since the merge of the PR for aarch64 Windows conflict...

Mizux commented:
```sh
[^v^]─corentinl@corentinl-macbookpro2 %CTEST_OUTPUT_ON_FAILURE=1 cmake --build build --target test -v
Running tests...
/opt/homebrew/Cellar/cmake/3.26.4/bin/ctest --force-new-ctest-process 
Test project /Users/corentinl/work/cpu_features/build
    Start 1: bit_utils_test
1/4 Test #1: bit_utils_test ...................   Passed    0.00 sec
    Start 2: string_view_test
2/4 Test #2: string_view_test .................   Passed    0.00 sec
    Start 3: stack_line_reader_test
3/4 Test #3: stack_line_reader_test ...........   Passed    0.00 sec
    Start 4: cpuinfo_aarch64_test
4/4 Test #4: cpuinfo_aarch64_test .............   Passed    0.65 sec

100% tests passed, 0 tests failed out of 4
```

```sh
[0]─[~/work/cpu_features]-[aarch64-darwin-support]
[^u^]─corentinl@corentinl-macbookpro2 %build/test/cpuinfo_aarch64_test 
Running main() from gmock_main.cc
[==========] Running 2 tests from 2 test suites.
[----------] Global test environment set-up.
[----------] 1 test from CpuinfoAarch64Test
[ RUN      ] CpuinfoAarch64Test.Aarch64FeaturesEnum
[       OK ] CpuinfoAarch64Test.Aarch64FeaturesEnum (0 ms)
[----------] 1 test from CpuinfoAarch64Test (0 ms total)

[----------] 1 test from CpuidAarch64Test
[ RUN      ] CpuidAarch64Test.FromDarwinSysctlFromName
[       OK ] CpuidAarch64Test.FromDarwinSysctlFromName (0 ms)
[----------] 1 test from CpuidAarch64Test (0 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 2 test suites ran. (0 ms total)
[  PASSED  ] 2 tests.
```

gchatelet commented:
@arkivm @Mizux 

Looking at https://github.com/google/cpu_features/pull/204/files#diff-fc9e9a184d90d6afa52754eeef66a041131b11a4fddc7cb8668f7cf54d6d2585R54-R75 I feel like there are still quite a few features that need to be filled in from `sysctlbyname`. Do we have a way to get a complete list of available options?

arkivm commented:
I rebased the patches and updated a few flags that got changed. There are a bunch of [additional features](https://gist.github.com/arkivm/c660fc9dbaaaff8d52f0c3c7e62da855) on aarch64 apple m2 silicon. Does those features have to match with the Linux's [hwcap](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm64/include/uapi/asm/hwcap.h)?

gchatelet commented:
> I rebased the patches and updated a few flags that got changed.

Thx !

> There are a bunch of [additional features](https://gist.github.com/arkivm/c660fc9dbaaaff8d52f0c3c7e62da855) on aarch64 apple m2 silicon. Does those features have to match with the Linux's [hwcap](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm64/include/uapi/asm/hwcap.h)?

AFAIU both `sysctlbyname` and `hwcap` are Linux maintained variables so I hope they are consistent with each other  🤷.
But for instance I don't quite know why we have `hw.optional.arm.FEAT_SHA256` vs `HWCAP_SHA2`...

Mizux commented:
sysctl on M1: https://gist.github.com/Mizux/fc3309b1efeb1454cc7b093c5f23992e
(ed `system_profiler SPHardwareDataType` seems cool to identify device model)

gchatelet commented:
Thx @Mizux, the list of features is the same between `M1` and `M2` but the following ones are `0` on `M1` and `1` on `M2`.

```
hw.optional.arm.FEAT_PAuth2
hw.optional.arm.FEAT_FPAC
hw.optional.arm.FEAT_BF16
hw.optional.arm.FEAT_I8MM
hw.optional.arm.FEAT_BTI
```
